[7806] plans and projects api caching

m4ra opened the pull request:
**Tasks**
- [x] PR name contains story or task reference
- [x] Documentation (docs and inline)
- [x] Tests (including n+1 and django_assert_num_queries where applicable)
- [x] Changelog

goapunk commented on meinberlin/config/settings/production.py:
```diff
@@ -1,6 +1,14 @@
+
+CACHES = {
+    "default": {
+        "BACKEND": "django.core.cache.backends.redis.RedisCache",
```
We'll also need to add this to `travis.py` otherwise it will use the `LocMemCache` from below

goapunk commented on meinberlin/apps/extprojects/api.py:
```diff
@@ -18,3 +20,13 @@ def get_queryset(self):
+        if data is None:
+            queryset = self.filter_queryset(self.get_queryset())
+            serializer = self.get_serializer(queryset, many=True)
+            cache.set("extprojects", serializer.data, 360)
```
One thing we discussed on the old PR: if a project starts at `12:00` and the cache is set at `11:59` the project won't appear in the list for `cache_time` -1 minutes which is not ideal. I think we considered using celery-beat to periodically invalidate the cache or setting the cache_time to the time of the earliest project which is about to start

m4ra replied:
Let's talk about it tmr after standup. For instance, a periodic task that checks whether a project starts in the next 5 mins? running every 5 mins? if yes. the task clears the cache for these type of projects?
@goapunk

goapunk reviewed (changes requested):
Some initial comments and the question about how we handle cache expiry times

m4ra commented on meinberlin/apps/projects/signals.py:
```diff
@@ -20,3 +20,5 @@ def post_save_delete(sender, instance, update_fields, **kwargs):
         ["projects_active", "projects_future", "projects_past", "private_projects"]
     )
+    # TODO: add redis key for earliest next project starts
+    # TODO: add redis key for earliest next project ends
```
@goapunk here I will fetch the earliest start and latest end date from phase objects and set them in redis db

m4ra commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,72 @@
+    if ends:
+        # remove redis key next_project_ends
+        redis_db.delete("next_project_ends")
+
```
@goapunk have a look and we can discuss it again TMR. The logic is to check every 10 min two things:
one task for future and one for past projects:  if the earliest future/latest past project is happening in the next 10 mins, and if yes then we call the actual reset cache task with a countdown. If not, then one of the next callings of the periodic tasks will do it.

We only query the DB when there is no redis key set, and when a project is saved.

goapunk replied:
structure looks good to me, left a small question but I think this will be a nice solution

m4ra commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,72 @@
+def schedule_reset_cache_for_future_projects():
+    # fetch earliest future project
+    # check if redis has a next_project_starts key
+    next_project_starts = redis_db.get("next_project_starts")
```
I thought to not mess with the cache, in case of cache.clear(). But prob doesn't matter, I will use cache.

m4ra commented:
@goapunk have another look when you can. I need to add a test, e.g adding future projects and check if the are added in the cache.

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,107 @@
+def reset_cache_for_projects(starts, ends):
+    if starts:
+        # remove redis key next_project_start
+        cache.delete("next_project_starts")
```
we still need to actually delete the cache here, right?

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,107 @@
+
+    if not next_project_starts:
+        # set the redis key: value
+        cache.set("next_project_starts", get_next_project_starts())
```
do we need to set `next_project_starts` here? otherwise it will be `None` below

m4ra replied:
line 57 does that, setting the 'next_project_starts'.

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,107 @@
+
+@shared_task
+def reset_cache_for_projects(starts, ends):
+    if starts:
```
Do we have to call schedule_reset_cache_for_projects_* here again or run this for all projects starting within 10 minutes ? Considering the following case:
1. Project A starts at 12:01, Project B starts at 12:03
2. The schedule task runs at 12:00 and now schedules a reset_cache_for_projects for 12:01
3. At 12:01 the cache is reset and now includes Project A
4. At 12:03 Project B won't be visible yet because schedule_reset only runs at 12:10
5. At 12:10 schedule_reset runs again, if there's no future project starting the cache won't be updated and Project B still won't be visible, if there is a future Project C, Project B will only become visible once Project C starts

This is more complicated then I hoped :/ Not sure how to make this bulletproof. Maybe we can have a look together?

m4ra replied:
Quite a niche case, but valid point nonetheless. I will check for a list of future projects that are to be published in the next 10 mins with the function `get_next_project_starts()` and apply the cache reset for that list of projects.

goapunk commented on meinberlin/apps/extprojects/signals.py:
```diff
@@ -0,0 +1,10 @@
+from .models import ExternalProject
+
+
+@receiver(post_save, sender=ExternalProject)
```
should we add `post_delete` here as well?

goapunk commented on meinberlin/apps/plans/signals.py:
```diff
@@ -0,0 +1,12 @@
+
+
+@receiver(post_save, sender=Plan)
+@receiver(pre_delete, sender=Plan)
```
should we use `post_delete` instead of `pre_delete? seems to be more logical

goapunk commented on meinberlin/apps/projects/signals.py:
```diff
@@ -0,0 +1,49 @@
+
+
+@receiver(post_save, sender=Project)
+@receiver(pre_delete, sender=Project)
```
`should we use post_delete instead of `pre_delete? seems to be more logical

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,169 @@
+def get_next_project_ends() -> str:
+    """
+    Helper function to query the db and
+    retreive the erliest phase that will end.
```
typo: erliest -> earliest

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,169 @@
+                "extprojects",
+            ]
+        )
+        if cache.get("projects_*") or cache.get("extprojects"):
```
`cache.get("projects_*")` will only work with redis, not with any of the other caching backends, right?

goapunk replied:
hmm, I can't find any information on this though. The django docs don't mention wildcards (or I'm too stupid to find them), only the django-redis package mentions it. However if we want to be on the safe side I think we can use `cache.get_many(['projects_activeParticipation', 'projects_pastParticipation', ... , 'extprojects',])`

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,169 @@
+        cache.delete("next_project_starts")
+        cache.delete_many(
+            [
+                "projects_activeParticipation",
```
above we use other keys:
```
  cache.delete_many(
        [
            "projects_active*",
            "projects_future*",
            "projects_past*",
            "private_projects",
            "extprojects",
        ]
    )
```
i assume we need the same here ? :)

goapunk commented on meinberlin/apps/projects/signals.py:
```diff
@@ -0,0 +1,49 @@
+    Capture any new phases that may got created/updated while saving a project.
+    """
+
+    cache.delete_many(
```
below and in the docs it says the keys are:
```
       cache.delete_many(
            [
                "projects_activeParticipation",
                "projects_futureParticipation",
                "private_projects",
                "extprojects",
            ]
        )
```
I assume you forgot to change it here ? :)

goapunk commented on meinberlin/apps/projects/signals.py:
```diff
@@ -0,0 +1,49 @@
+def post_dashboard_signal_delete(sender, project, user, **kwargs):
+    cache.delete_many(
+        [
+            "projects_active*",
```
do we need the globbing here? the wildcard will only match `projects_activeParticipation`, right ? And the wildcard will make it incompatible with non-redis cache backends

goapunk commented on meinberlin/config/settings/production.py:
```diff
@@ -30,3 +39,16 @@
+CELERY_BEAT_SCHEDULE = {
+    "update-future-projects-every-10-mim": {
+        "task": "schedule_reset_cache_for_projects_becoming_active",
+        "schedule": crontab(minute="*/10"),
```
`crontab` is missing an import here : `from celery.schedules import crontab`

goapunk reviewed (commented):
Nice, looks good to me so far, I left some comments with questions. I'm not sure about the failing tests, seems to be a bit tricky with the combination of redis + `CELERY_TASK_ALWAYS_EAGER=True`. Without it we need a running celery worker though

m4ra commented:
@goapunk there was a typo in the reset cache task of the key to be deleted. As I was testing locally and because of the asterisks I was using, something was not captured. Anyhow I simplified the tests and removed the API calls as these are part of the `test_api_caching`. Also I ended up using a 10min check and a list of projects that will become past same as for future projects, otherwise we may miss any other expiring between the 10mins if one is already set for expiration. Looking again at the tasks and functions I wonder if they should be consolidate instead of having separated for future and past projects.

goapunk commented:
> @goapunk there was a typo in the reset cache task of the key to be deleted. As I was testing locally and because of the asterisks I was using, something was not captured. Anyhow I simplified the tests and removed the API calls as these are part of the `test_api_caching`. Also I ended up using a 10min check and a list of projects that will become past same as for future projects, otherwise we may miss any other expiring between the 10mins if one is already set for expiration. Looking again at the tasks and functions I wonder if they should be consolidate instead of having separated for future and past projects.

I guess the only difference is the deletion of `projects_futureParticipation` vs `projects_pastParticipation`, so yeah, maybe we just delete them both and consolidate the code? What do you think?

m4ra commented:
@goapunk I merged the two scheduled tasks, the actual resets for future/past projects i left as it is, since it will save us from queering either future or past projects again via the url, which are time/power consuming.
I added also django query checks in the tests, very helpful and eye opening, as without cache, the api endpoint queries 12 times (with the prefetch-related).

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,192 @@
+        if cache.get_many(
+            [
+                "projects_activeParticipation",
+                "projects_futureParticipation" "private_projects",
```
is there a comma  + newline missing between the two values?

goapunk commented on meinberlin/apps/projects/tasks.py:
```diff
@@ -0,0 +1,192 @@
+        if cache.get_many(
+            [
+                "projects_activeParticipation",
+                "projects_pastParticipation" "private_projects",
```
is there a comma + newline missing between the two values?

goapunk commented on docs/api_caching.md:
```diff
@@ -0,0 +1,63 @@
+- schedule_reset_cache_for_projects()
+
+In case of projects becoming active the cache is cleared for:
+- `projects`
```
we don't clear this key anywhere, right?

goapunk commented on docs/api_caching.md:
```diff
@@ -0,0 +1,63 @@
+- `extprojects`
+
+in case of projects becoming past the cache is cleared for:
+- `projects`
```
we don't clear this key anywhere, right?

goapunk reviewed (changes requested):
Really nice! Just some small comments left

goapunk commented:
> @goapunk I merged the two scheduled tasks, the actual resets for future/past projects i left as it is, since it will save us from queering either future or past projects again via the url, which are time/power consuming. I added also django query checks in the tests, very helpful and eye opening, as without cache, the api endpoint queries 12 times (with the prefetch-related).

@m4ra yeah, the query check is really helpful, we definitely need to work on improving the queries/structure. Ideally we will have checks in all our tests in the future :crossed_fingers:

goapunk reviewed (approved):
Really cool :) Before we deploy it we need to make some changes in the admin repo
